File: Data Structures & Algorithms/house-robber/submission-5.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        left = mid = right = 0
        for i in range(len(nums) - 1, -1, -1):
            left, mid, right = (
                nums[i] + max(mid, right),
                left,
                mid
            )
        return max(left, mid)


# dp[i] = nums[i] + max(dp[i + 2], dp[i + 3])

# nums = [1, 1, 3, 3]

# dp = [?, ?, ?, ?, 0, 0, 0]

    #             (0)
    #             r:4
    #     (2)             (3)
    #     r:3             r:3
    # (4)     (5)     (5)     (6)
    # r:0     r:0     r:0     r:0

File: Data Structures & Algorithms/binary-tree-diameter/submission-5.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
private:
    int dfs(TreeNode* node, int& max_diameter) {
        if (node == nullptr) return 0;

        int depth_left = dfs(node->left, max_diameter);
        int depth_right = dfs(node->right, max_diameter);

        max_diameter = max(max_diameter, depth_left + depth_right);

        return 1 + max(depth_left, depth_right);
    }

public:
    int diameterOfBinaryTree(TreeNode* root) {
        int max_diameter = 0;
        dfs(root, max_diameter);
        return max_diameter;
    }
};
